src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,219 @@
+
+        next_execute_time: int
+
+        if periodic_repair_interval is None:
+            periodic_repair_interval = sys.maxsize
+        if periodic_deploy_interval is None:
+            periodic_deploy_interval = sys.maxsize
+
+        if is_dirty:
+            next_execute_time = min(periodic_deploy_interval, periodic_repair_interval)
+        else:
+            next_execute_time = periodic_repair_interval
+
+        if next_execute_time == sys.maxsize:
+            return
+
+        # We have to schedule the next deploy sooner than previously though:
+        # Cancel previous schedule and re-schedule
+        if next_execute_time < self.time_to_next_deploy:
+            self.cancel_next_schedule()
+
+        # A deploy is already scheduled for this resource
+        if self.next_schedule_handle is not None:
+            return
+
+        def _create_repair_task() -> None:
+            self.repair_handle = asyncio.create_task(action_function(self.resource, next_execute_time))
+
+        self.time_to_next_deploy = next_execute_time
+        self.next_schedule_handle = asyncio.get_running_loop().call_later(next_execute_time, _create_repair_task)
+
+    def cancel_next_schedule(self) -> None:
+        """
+        Cancel the callback that schedules the next repair task
+        """
+        if self.next_schedule_handle is not None:
+            self.next_schedule_handle.cancel()
+
+    def cancel_repair_task(self) -> None:
+        """
+        Cancel the underlying repair task itself
+        """
+        if self.repair_handle is not None:
+            self.repair_handle.cancel()
+
+
+class TimerManager:
+    def __init__(self, resource_scheduler: "inmanta.deploy.scheduler.ResourceScheduler"):
+        """
+        :param resource_scheduler: Back reference to the ResourceScheduler that was responsible for
+            spawning this TimerManager.
+        """
+        self.resource_timers: dict[ResourceIdStr, ResourceTimer] = {}
+
+        self.global_periodic_repair_task: ScheduledTask | None = None
+        self.global_periodic_deploy_task: ScheduledTask | None = None
+
+        self.periodic_repair_interval: int | None = None
+        self.periodic_deploy_interval: int | None = None
+
+        self._resource_scheduler = resource_scheduler
+
+        self._sched = Scheduler("Resource scheduler")
+
+    def reset(self) -> None:
+        self.stop()
+
+        self.global_periodic_repair_task = None
+        self.global_periodic_deploy_task = None
+        self.periodic_repair_interval = None
+        self.periodic_deploy_interval = None
+
+    def stop(self) -> None:
+        for timer in self.resource_timers.values():
+            timer.cancel_next_schedule()
+
+        for task in [self.global_periodic_repair_task, self.global_periodic_deploy_task]:
+            if task:
+                self._sched.remove(task)
+
+    def join(self) -> None:
+        for timer in self.resource_timers.values():
+            timer.cancel_repair_task()
```

Comment:
join is expected to wait for the cancel to be done, which is not what this method does. (It is not async, so it clearly doesn't wait)

Reply:
Refactored this a bit to the following:

Clarify in the `ResourceTimer` class between:
```
        # "Inner" asyncio task responsible for queuing a Deploy task
        self.trigger_deploy: asyncio.Task[None] | None = None
        # "Outer" handle to the soon-to-be-called trigger_deploy
        self.next_schedule_handle: asyncio.TimerHandle | None = None
```
        
- When `timer_manager` is stopped: 
  - cancel `next_schedule_handle` for each timer
- When `timer_manager` is joined:
  - await the `trigger_deploy` if any for each timer (i.e. wait for the enqueuing to be done)